README.md
TwitX
================================================================================

Load and display Twitter feeds and post Tweets using the Twitter 1.1 REST API

Features:
--------------------------------------------------------------------------------
This MODX Evolution snippet loads Twitter feeds using the Twitter 1.1 REST API. You will need to create a Twitter app and get the keys, secrets and token on https://dev.twitter.com/apps/new

Installation:
--------------------------------------------------------------------------------
1. Upload all files into the new folder *assets/snippets/twitx*
2. Create a new snippet called TwitX with the following snippet code
    `<?php return include (MODX_BASE_PATH . 'assets/snippets/twitx/twitx.snippet.php'); ?>`
3. If you are using the PHx plugin (or more than one Snippet/Plugin with evoChunkie class) you should copy the PHx modifier from the folder `phx_modifier` to `assets/plugins/phx/modifiers`
4. Create a Twitter app on https://dev.twitter.com/apps/new and note the Consumer Key and the Consumer Secret. On that page you have to create an access token and note the Access Token and the Access Token Secret. These values have to be used later in the snippet call.

Usage
--------------------------------------------------------------------------------

Basic snippet call:

```
[!TwitX?
    &twitter_consumer_key=`aaaa`
    &twitter_consumer_secret=`bbbb`
    &twitter_access_token=`cccc`
    &twitter_access_token_secret=`dddd`
!]
```
The snippet could use the following properties:

Property | Description | Default
---- | ----------- | -------
twitter_consumer_key | **required:** your twitter consumer token  | -
twitter_consumer_secret | **required:** your twitter consumer secret | -
twitter_access_token | **required:** your twitter access token | -
twitter_access_token_secret | **required:** your twitter access token secret | -
limit | limit the number of statuses to display | 5
twitTpl | template chunk for one twitter status | `twitTpl.html` in folder `templates`
timeline | name of the twitter timeline* | user_timeline
decodeUrls | decode shortened t.co urls | 1;
cache | seconds the twitter feed ist cached (0 = disable cache) | 7200
screen_name | screen name of the user the twitter feed is loaded from | -
targetBlank | if not set to 0 all a tags have target="_blank" | 1
relNofollow | if not set to 0 all a tags have rel="nofollow" | 0
include_rts | include retweets | 1
outputSeparator | separator between two twitTpl elements | `newline`
toPlaceholder | a placeholder name the snippet output is assigned to. surpesses normal snippet output | -

* Possible timelines: favorites, mentions_timeline, user_timeline, home_timeline, retweets_of_me

Placeholder
--------------------------------------------------------------------------------
Look in the Twitter 1.1 REST API for possible placeholder in the TwitX template.

Each member of the response object could be referenced by its name. Members of a member could be referenced by name of the parent member and the name of the member divided by `.`.

Example response object:

```
[
  {
    "created_at": "Wed Aug 29 17:12:58 +0000 2012",
    "entities": {
      "urls": [
        {
          "expanded_url": "https://dev.twitter.com/blog/twitter-certified-products"
        }
      ]
    }
  }
]
```

Possible placeholder showing a value are: `[+created_at+]` `[+entities.urls.expanded_url+]`

These placeholder could be modified by PHx modifier.

Notes:
--------------------------------------------------------------------------------
1. Uses Cache Class: http://www.axel-hahn.de/php_contentcache.php
2. Uses TwitterOAuth Class: https://github.com/abraham/twitteroauth
3. TwitX is an Evolution Port of TwitterX: http://modx.com/extras/package/twitterx
4. TwitX uses its own cache settings and cache files and its cache could only be bypassed before the default or presetted cache lifetime by using the parameter `` &cache=`0` ``. During the testing phase this cache setting should be used.


HISTORY.md
History
================================================================================

- 1.2 November 5, 2013
    - install set for http://extras.evolution-cms.com/
- 1.1 June 15, 2013
    - Updated twitterOAuth class (to solve Twitter 1.1 REST API problems)
	- Changed link replacement in Tweet output
    - Removed twitx_format PHx modifier
    - New parameter targetBlank and relNofollow
- 1.0b June 3, 2013
	- Send Tweets
    - All fields in response from Twitter 1.1 REST API are now usable as placeholder in TwitX templates
    - New decodeUrls parameter - defaults to 1
- 0.9.1 May 24m 2013
    - Bugfix for evoChunkie
    - Updated documentation
- 0.9 February 12, 2013
    - Initial public release


assets/snippets/twitx/twitx.snippet.php
<?
/**
 * TwitX
 *
 * This MODX snippet loads Twitter feeds using the Twitter 1.1 REST API.
 * You will need to create a Twitter app and get the keys and tokens:
 * https://dev.twitter.com/apps/new
 *
 * This Snippet uses evoChunkie
 * This Snippet uses TwitterOAuth: https://github.com/abraham/twitteroauth
 * This Snippet uses Cache Class: http://www.axel-hahn.de/php_contentcache.php
 *
 * TwitX is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * TwitX is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * MODX Evolution Port of TwitterX
 * @author Thomas Jakobi <thomas.jakobi@partout.info>
 * @version 1.2
 *
 * description: <strong>1.2</strong> Load and display Twitter feeds and post Tweets using the Twitter 1.1 REST API
 *
 * TwitterX author: Stewart Orr @ Qodo Ltd <stewart@qodo.co.uk>
 */
// Twitter API keys and secrets
$twitter_consumer_key = isset($twitter_consumer_key) ? $twitter_consumer_key : FALSE;
$twitter_consumer_secret = isset($twitter_consumer_secret) ? $twitter_consumer_secret : FALSE;
$twitter_access_token = isset($twitter_access_token) ? $twitter_access_token : FALSE;
$twitter_access_token_secret = isset($twitter_access_token_secret) ? $twitter_access_token_secret : FALSE;

// Other options
$mode = isset($mode) ? $mode : 'timeline';
$limit = isset($limit) ? $limit : 5;
$twitTpl = isset($twitTpl) ? $twitTpl : '@FILE:assets/snippets/twitx/templates/twitTpl.html';
$tweetedTpl = isset($tweetedTpl) ? $tweetedTpl : '@FILE:assets/snippets/twitx/templates/tweetedTpl.html';
$timeline = isset($timeline) ? $timeline : 'user_timeline';
$decodeUrls = isset($decodeUrls) ? (boolean) $decodeUrls : TRUE;
$cache = isset($cache) ? $cache : 7200;
$screen_name = isset($screen_name) ? $screen_name : '';
$tweet = isset($tweet) ? $tweet : '';
$targetBlank = (isset($targetBlank) && $targetBlank == '0') ? '' : ' target="_blank"';
$relNofollow = (isset($relNofollow) && $relNofollow == '0') ? '' : ' rel="nofollow"';
$include_rts = (isset($include_rts) && (!$include_rts)) ? 0 : 1;
$outputSeparator = isset($outputSeparator) ? $outputSeparator : "\r\n";
$toPlaceholder = isset($toPlaceholder) ? $toPlaceholder : '';

if (!class_exists('evoChunkie')) {
	require MODX_BASE_PATH . 'assets/snippets/twitx/includes/chunkie/chunkie.class.inc.php';
}
if (!class_exists('evoCache')) {
	require MODX_BASE_PATH . 'assets/snippets/twitx/includes/cache/cache.class.php';
}
if (!class_exists('TwitterOAuth')) {
	require MODX_BASE_PATH . 'assets/snippets/twitx/includes/twitteroauth/twitteroauth.php';
}

if (!function_exists('twitxFormat')) {

	function twitxFormat($entry, $targetBlank, $relNofollow) {
		//replace hashtags,
		if (count($entry['entities']['hashtags'])) {
			foreach ($entry['entities']['hashtags'] as $entity) {
				$entry['text'] = str_replace('#' . $entity['text'], '<a href="https://twitter.com/search?q=%23' . $entity['text'] . '&src=hash"' . $targetBlank . $relNofollow . '>#' . $entity['text'] . '</a>', $entry['text']);
			}
		}
		// urls,
		if (count($entry['entities']['urls'])) {
			foreach ($entry['entities']['urls'] as $entity) {
				$entry['text'] = str_replace($entity['url'], '<a href="' . $entity['expanded_url'] . '"' . $targetBlank . $relNofollow . '>' . $entity['display_url'] . '</a>', $entry['text']);
			}
		}
		// user mentions
		if (count($entry['entities']['user_mentions'])) {
			foreach ($entry['entities']['user_mentions'] as $entity) {
				$entry['text'] = str_replace('@' . $entity['screen_name'], '<a href="https://twitter.com/' . $entity['screen_name'] . '"' . $targetBlank . $relNofollow . '>@' . $entity['screen_name'] . '</a>', $entry['text']);
			}
		}
		// and media entities
		if (count($entry['entities']['media'])) {
			foreach ($entry['entities']['media'] as $entity) {
				$entry['text'] = str_replace($entity['url'], '<a href="' . $entity['expanded_url'] . '"' . $targetBlank . $relNofollow . '>' . $entity['display_url'] . '</a>', $entry['text']);
			}
		}
		return $entry;
	}

}

// HTML output
$output = array();
// If they haven't specified the required Twitter keys, we cannot continue...
if (!$twitter_consumer_key || !$twitter_consumer_secret || !$twitter_access_token || !$twitter_access_token_secret) {
	$output[] = '<strong>TwitX Error:</strong> Could not load TwitX as required values were not passed.';
} else {
	// Test for required function(s)
	if (!function_exists('curl_init')) {
		$output[] = "<strong>TwitX Error:</strong> cURL functions do not exist, cannot continue.";
	} else {
		switch ($mode) {
			case 'tweet':
				if ($screen_name == '') {
					$output[] = "<strong>TwitX Error:</strong> No Twitter screen name set for tweeting.";
					break;
				}

				// Create new twitteroauth
				$twitteroauth = new TwitterOAuth($twitter_consumer_key, $twitter_consumer_secret, $twitter_access_token, $twitter_access_token_secret);

				// We want to use JSON format
				$twitteroauth->format = 'json';
				$twitteroauth->decode_json = FALSE;

				$options = array(
					'screen_name' => $screen_name,
					'text' => urlencode(substr($modx->stripTags($tweet), 0, 140))
				);
				$json = $twitteroauth->post('direct_messages/new.json', $options);
				$json = json_decode($json, TRUE);

				if (isset($json['error'])) {
					$output[] = "<strong>TwitX Error:</strong> Could not send the Tweet. Twitter responded with the error '" . $json->error . "'.";
				} else {
					$parser = new evoChunkie($tweetedTpl);
					$parser->CreateVars($json);
					$output[] = $parser->Render();
				}

				break;
			case 'timeline':
			default:
				// Try loading the data from cache first
				$myCache = new evoCache('TwitX', $screen_name . '_' . $timeline);

				if ($myCache->isExpired() || $cache == 0) {
					// Load the TwitterOAuth lib required if not exists
					// Create new twitteroauth
					$twitteroauth = new TwitterOAuth($twitter_consumer_key, $twitter_consumer_secret, $twitter_access_token, $twitter_access_token_secret);

					// We want to use JSON format
					$twitteroauth->format = 'json';
					$twitteroauth->decode_json = FALSE;

					// Request statuses with optinal parameters
					$options = array(
						'count' => $limit,
						'include_rts' => $include_rts
					);
					// If we have a screen_name, pass this to Twitter API
					if ($screen_name != '') {
						$options['screen_name'] = $screen_name;
					}

					if ($decodeUrls) {
						$options['include_entities'] = true;
					}

					// If we are viewing favourites or regular statuses
					if ($timeline != 'favorites') {
						$timeline = 'statuses/' . $timeline;
					}
					$json = $twitteroauth->get($timeline, $options);

					// No errors? Save to Cache
					$status = json_decode($json);
					if (!isset($status->error)) {
						$myCache->write($json, $cache);
					}
				} else {
					// read cached data
					$json = $myCache->read();
				}

				// Decode this now that we have used it above in the cache
				$json = json_decode($json, TRUE);

				// If there any errors from Twitter, output them...
				if (isset($json['error'])) {
					$output[] = "<strong>TwitX Error:</strong> Could not load TwitX. Twitter responded with the error '" . $json->error . "'.";
				} else {
					// For each result, output it
					foreach ($json as &$j) {
						$parser = new evoChunkie($twitTpl);
						if ($decodeUrls) {
							// work retweeted text
							if (isset($j['retweeted_status'])) {
								$j['retweeted_status'] = twitxFormat($j['retweeted_status'], $targetBlank, $relNofollow);
							}
							// work text
							$j = twitxFormat($j, $targetBlank, $relNofollow);
						}
						$parser->CreateVars($j);
						// Parse chunk passing values
						$output[] = $parser->Render();
					}
				}

				// Added option to output to placeholder
				if ($toPlaceholder != '') {
					$modx->setPlaceholder($toPlaceholder, implode($outputSeparator, $output));
					$output = array();
					$outputSeparator = '';
				}
		}
	}
}
return implode($outputSeparator, $output);
?>